### PR TITLE
feat: as a portal creator, I can publish my Portal SC

### LINKED ISSUE #60: [US] As a Portal Creator, I can publish my Portal SC
As a Portal, I want to publish my Portal on the Portal Registry so that they can be used by Attestation Issuers
More details on this specification in the issue #7 

## Requirements
* Anyone can create a Portal as this is a SC. No rule to prevent anyone from creating a Portal
* Portals are registered through the `PortalRegistry` smart contract
* Portals are identified via their Smart Contract addresses (i.e portalId = portal SC address)
* Portals published should be stored in an array (for discoverability)
* Portals' metadata should be stored in a mapping when the module is published
    * The Portal smart contract address is used as a key
    * Portal name and description are stored inside the mapping (struct)
* Publishing a Portal should trigger an event

## Todo
- [ ] Create an TransparentUpgradeable `PortalRegistry` contract
- [ ] Create a `Portal` struct to store a Portal metadata: `address id, string name, string description`
- [ ] Add a mapping to store Portals (`address => Portal`)
- [ ] Add an array to store modules
- [ ] Add a `register(address, string, description)` function to add a new Portal to the mapping and the array
- [ ] To register a Portal, it must have a name and an address non empty, throw custom errors if not
- [ ] Throw custom error on duplicate
- [ ] When a Portal is registered, emit a custom `PortalRegistered` event
- [ ] Add a `PortalRegistry` script to be able to deploy it easily + a pnpm command to trigger it
- [ ] Add a `PortalRegistry` set of unit tests to cover the happy path + the errors and events triggering

## Acceptance criteria
* Should include a script to deploy the `PortalRegistry`
* Should include 100% unit-tests (including accessing the metadata through the mapping)

## Additional information
| Workload estimate | 
|--------|
| TBD |

### PR DESCRIPTION
**Pull request by 0xEillo**
### Related ticket

Fixes #60 

### Type of change

- [ ] Chore
- [ ] Bug fix
- [x] New feature
- [ ] Documentation update

## Check list

- [x] Unit tests for any smart contract change
- [x] Contracts and functions are documented

### CODE REVIEW

File: src/PortalRegistry.sol [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  event PortalRegistered(string name, string description, address moduleAddress);
+
+  /**
+   * @notice Contract initialization
+   */
+  function initialize() public initializer {}
+
+  /**
+   * @notice Registers a Portal to the PortalRegistry
+   * @param id the module name
```
Comment (satyajeetkolhapure) [inline src/PortalRegistry.sol:41]: portal instead of module

File: test/PortalRegistry.t.sol [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+  function test_isRegistered() public {
+    assertEq(portalRegistry.isRegistered(address(validPortal)), false);
+    portalRegistry.register(address(validPortal), expectedName, expectedDescription);
+    assertEq(portalRegistry.isRegistered(address(validPortal)), true);
+  }
+}
+
+contract ValidPortal is IPortal, IERC165 {
+  function attest(bytes32 /*schemaId*/, bytes memory /*attestationData*/) external view returns (bool) {
```
Comment (satyajeetkolhapure) [inline test/PortalRegistry.t.sol:91]: I am not sure if this can be pure instead of view. Compiler warning only.

File: test/PortalRegistry.t.sol [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    return true;
+  }
+
+  function getModules() external view returns (address[] memory) {
```
Comment (satyajeetkolhapure) [inline test/PortalRegistry.t.sol:95]: I am not sure if this can be pure instead of view. Compiler warning only.

**Comment by satyajeetkolhapure** [comment]
For the coverage, dev branch has a fix in smart-contracts-test.yml to not include test folder.

File: src/PortalRegistry.sol [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  event PortalRegistered(string name, string description, address moduleAddress);
+
+  /**
+   * @notice Contract initialization
+   */
+  function initialize() public initializer {}
+
+  /**
+   * @notice Registers a Portal to the PortalRegistry
+   * @param id the portal name
```
Comment (orbmis) [inline src/PortalRegistry.sol:41]: Are these code comments correct? Should `id` actually be described as "portal contract address and unique identifier" or something

File: src/PortalRegistry.sol [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    // Emit event
+    emit PortalRegistered(name, description, id);
+  }
+
+  /**
+   * @notice Get a Portal by its address
+   * @param id The address of the Portal
+   * @return The Portal
+   */
+  function getPortals(address id) public view returns (Portal memory) {
```
Comment (orbmis) [inline src/PortalRegistry.sol:78]: nitpick: Should this be `getPortaByAddress`?

**Review by orbmis — APPROVED** [review]

**Review by satyajeetkolhapure — APPROVED** [review]

**Pull request merged**